File: android/app/src/main/kotlin/org/gaindrive/android/ui/components/ServerBadge.kt
package org.gaindrive.android.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

/**
 * Which server a row came from.
 *
 * Deliberately quiet — a tinted label rather than a chip. It appears on every
 * row in merged scope, so anything louder would compete with the content it is
 * annotating. Callers pass null in single-server scope, and nothing renders.
 */
@Composable
fun ServerBadge(name: String?) {
	if (name == null) return
	Text(
		text = name,
		style = MaterialTheme.typography.labelSmall,
		color = MaterialTheme.colorScheme.onSurfaceVariant,
		maxLines = 1,
		overflow = TextOverflow.Ellipsis,
		modifier = Modifier
			.clip(RoundedCornerShape(4.dp))
			.background(MaterialTheme.colorScheme.surfaceVariant)
			.padding(horizontal = 6.dp, vertical = 2.dp),
	)
}

/** Several of them, for a row that stands for the same artist on each. */
@Composable
fun ServerBadges(names: List<String>) {
	if (names.size < 2) {
		ServerBadge(names.firstOrNull())
		return
	}
	Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
		names.forEach { ServerBadge(it) }
	}
}

File: ios/GainDrive/UI/Components/ServerBadge.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import SwiftUI

/// Which server a row came from.
///
/// Deliberately quiet - a tinted label rather than a chip. It appears on every
/// row in merged scope, so anything louder would compete with the content it is
/// annotating.
///
/// Callers pass `nil` in single-server scope and nothing renders, which is what
/// keeps the app looking like a single-server client until it isn't one.
struct ServerBadge: View {
	let name: String?

	var body: some View {
		if let name, !name.isEmpty {
			Text(name)
				.font(.caption2)
				.lineLimit(1)
				.padding(.horizontal, 6)
				.padding(.vertical, 2)
				.background(.quaternary, in: RoundedRectangle(cornerRadius: 4, style: .continuous))
				.foregroundStyle(.secondary)
				// A bare server name read out in the middle of a row says
				// nothing about what it is; the row is announced as one
				// utterance, so this has to carry its own preposition.
				.accessibilityLabel("on \(name)")
		}
	}
}

/// Plural, because a row whose duplicates were collapsed stands for every
/// server that has the album - and hiding that would make the missing second
/// row look like a bug.
struct ServerBadges: View {
	let names: [String]

	var body: some View {
		if !names.isEmpty {
			HStack(spacing: 4) {
				ForEach(names, id: \.self) { ServerBadge(name: $0) }
			}
		}
	}
}
